benchmarks/ddp/benchmark.py

```diff
@@ -0,0 +1,276 @@
+#!/usr/bin/env python3
+#
+# Measure distributed training iteration time.
+#
+# This program performs a sweep over a) a number of model architectures, and
+# b) an increasing number of processes. This produces a 1-GPU baseline,
+# an 8-GPU baseline (if applicable), as well as measurements for however
+# many processes can participate in training.
+#
+
+import argparse
+import gc
+import io
+import itertools
+import json
+import os
+import shlex
+import subprocess
+import sys
+import time
+
+import numpy as np
+import torch
+import torch.distributed as dist
+import torch.nn as nn
+import torch.optim as optim
+import torchvision
+
+
+def allgather_object(obj):
+    buffer = io.BytesIO()
+    torch.save(obj, buffer)
+    input_tensor = torch.ByteTensor(list(buffer.getvalue()))
+    input_length = torch.IntTensor([input_tensor.size(0)])
+    dist.all_reduce(input_length, op=dist.ReduceOp.MAX)
+    input_tensor.resize_(input_length[0])
+    output_tensors = [
+        torch.empty(input_tensor.size(), dtype=torch.uint8)
+        for _ in range(dist.get_world_size())
+    ]
+    dist.all_gather(output_tensors, input_tensor)
+    output = []
+    for tensor in output_tensors:
+        buffer = io.BytesIO(np.asarray(tensor).tobytes())
+        output.append(torch.load(buffer))
+    return output
+
+
+def allgather_run(cmd):
+    proc = subprocess.run(shlex.split(cmd), capture_output=True)
+    assert(proc.returncode == 0)
+    return allgather_object(proc.stdout.decode("utf-8"))
+
+
+
+def allequal(iterator):
+    iterator = iter(iterator)
+    try:
+        first = next(iterator)
+    except StopIteration:
+        return True
+    return all(first == rest for rest in iterator)
+
+
+def benchmark_process_group(pg, benchmark):
+    torch.manual_seed(pg.rank())
+    torch.cuda.manual_seed(pg.rank())
+
+    model = benchmark.create_model()
+    data = [(benchmark.generate_inputs(), benchmark.generate_target())]
+    criterion = nn.CrossEntropyLoss()
+    optimizer = optim.SGD(model.parameters(),
+                          0.001,
+                          momentum=0.9,
+                          weight_decay=1e-4)
+    if pg.size() > 1:
```

Comment:
Is it possible to use DDP even for pg.size() == 1? This way all the results exercise the same code path.

Reply:
It is possible (I tried it). But the result is surprisingly off by a couple percent, indicating there is non-trivial overhead in DDP itself. I'm adding another option that adds a test run for 1 GPU with/without DDP, so that we always measure the impact. Already a good find!